pkg/vault/test/foundry/utils/BaseExtremeAmountsTest.sol

```diff
@@ -0,0 +1,546 @@
+        uint256[2] memory balancesRaw,
+        uint256 swapFee
+    ) public {
+        _setUpVariables(_boundBalances(balancesRaw), _boundSwapFee(swapFee));
+
+        _testAddLiquiditySingleTokenExactOutAndRemoveExactOut(
+            _boundExactBPTAmountForSingleTokenOperations(exactBPTAmount)
+        );
+    }
+
+    function testAddLiquiditySingleTokenExactOutAndRemoveExactOut_FuzzBPTAmount(uint256 exactBPTAmount) public {
+        _setUpVariables([HUGE_INIT_AMOUNT, HUGE_INIT_AMOUNT].toMemoryArray(), maxSwapFee / 2);
+
+        _testAddLiquiditySingleTokenExactOutAndRemoveExactOut(
+            _boundExactBPTAmountForSingleTokenOperations(exactBPTAmount)
+        );
+    }
+
+    function testAddLiquiditySingleTokenExactOutAndRemoveExactOut_FuzzSwapFee(uint256 swapFee) public {
+        swapFee = _boundSwapFee(swapFee);
+        _setUpVariables([HUGE_INIT_AMOUNT, HUGE_INIT_AMOUNT].toMemoryArray(), swapFee);
+
+        _testAddLiquiditySingleTokenExactOutAndRemoveExactOut(MIN_AMOUNT);
+    }
+
+    function _testAddLiquiditySingleTokenExactOutAndRemoveExactOut(uint256 exactBPTAmount) private {
+        vault.forceUnlock();
+
+        (uint256[] memory amountsIn, uint256 bptAmountOut, ) = vault.addLiquidity(
+            AddLiquidityParams({
+                pool: pool,
+                to: address(this),
+                maxAmountsIn: [0, maxAdditionalAmountsIn[1]].toMemoryArray(),
+                minBptAmountOut: exactBPTAmount,
+                kind: AddLiquidityKind.SINGLE_TOKEN_EXACT_OUT,
+                userData: bytes("")
+            })
+        );
+
+        assertEq(bptAmountOut, exactBPTAmount, "bpAmountOut should be equal to exactBPTAmount");
+
+        try
+            vault.removeLiquidity(
+                RemoveLiquidityParams({
+                    pool: pool,
+                    from: address(this),
+                    maxBptAmountIn: MAX_UINT128,
+                    minAmountsOut: [0, amountsIn[1]].toMemoryArray(),
+                    kind: RemoveLiquidityKind.SINGLE_TOKEN_EXACT_OUT,
+                    userData: bytes("")
+                })
+            )
+        returns (uint256 bptAmountIn, uint256[] memory amountsOut, bytes memory) {
+            assertEq(bptAmountIn, bptAmountOut, "bptAmountIn should be equal to bptAmountOut");
+            assertEq(amountsOut[0], amountsIn[0], "amountsOut[0] should be equal to amountsIn[0]");
+            assertEq(amountsOut[1], amountsIn[1], "amountsOut[1] should be equal to amountsIn[1]");
+        } catch (bytes memory reason) {
+            (bytes4 selector, bytes memory reasonWithoutSelector) = _extractSelectorFromErrorReason(reason);
+            (, uint256 balance, uint256 needed) = abi.decode(reasonWithoutSelector, (address, uint256, uint256));
+
+            assertEq(
+                selector,
+                IERC20Errors.ERC20InsufficientBalance.selector,
+                "Selector should be ERC20InsufficientBalance"
+            );
```

Comment:
I understand the problem here is that you're trying to burn more BPT than what you have. If I'm reading this correctly, you can exit with the exact same amounts you've entered, or the `remove` step reverts when you try to burn more BPT than what you have.

Instead of doing this, I'd use `bptAmountOut` as `maxBptAmountIn` in the remove step.

Maybe you can first run a query. If the query says that you will need more BPT in than what you have, run the actual operation using `expectRevert` and `bptAmountOut` as the limit; that will give you a known revert reason. If the query says that it's exactly the same, run the actual operation with the assertions inside the `try` block: 

```
            assertEq(bptAmountIn, bptAmountOut, "bptAmountIn should be equal to bptAmountOut");
            assertEq(amountsOut[0], amountsIn[0], "amountsOut[0] should be equal to amountsIn[0]");
            assertEq(amountsOut[1], amountsIn[1], "amountsOut[1] should be equal to amountsIn[1]");
```

That way you should avoid try / catch. In general "tests that might revert but it's not trivial to know when" are tricky, but I'd rather avoid try / catch logic here.